x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java

```diff
@@ -278,9 +293,174 @@ private void maybeUpdateMappings(Client localClient, Client remoteClient, Index
         }
     }
 
-    private void closeSession(Client remoteClient, DiscoveryNode node, String sessionUUID) {
-        ClearCcrRestoreSessionRequest clearRequest = new ClearCcrRestoreSessionRequest(sessionUUID, node);
-        ClearCcrRestoreSessionAction.ClearCcrRestoreSessionResponse response =
-            remoteClient.execute(ClearCcrRestoreSessionAction.INSTANCE, clearRequest).actionGet();
+    private static class RestoreSession implements Closeable {
+
+        private final int BUFFER_SIZE = 1 << 16;
+
+        private final Client remoteClient;
+        private final String sessionUUID;
+        private final DiscoveryNode node;
+        private final IndexShard indexShard;
+        private final ShardId shardId;
+        private final Store store;
+        private final RecoveryState recoveryState;
+        private final Store.MetadataSnapshot sourceMetaData;
+
+        RestoreSession(Client remoteClient, String sessionUUID, DiscoveryNode node, IndexShard indexShard, ShardId shardId,
+                       RecoveryState recoveryState, Store.MetadataSnapshot sourceMetaData) {
+            this.remoteClient = remoteClient;
+            this.sessionUUID = sessionUUID;
+            this.node = node;
+            this.indexShard = indexShard;
+            this.shardId = shardId;
+            this.store = indexShard.store();
+            this.store.incRef();
+            this.recoveryState = recoveryState;
+            this.sourceMetaData = sourceMetaData;
+        }
+
+        static RestoreSession openSession(Client remoteClient, ShardId leaderShardId, IndexShard indexShard, ShardId shardId,
+                                          RecoveryState recoveryState) {
+            String sessionUUID = UUIDs.randomBase64UUID();
+            PutCcrRestoreSessionAction.PutCcrRestoreSessionResponse response = remoteClient.execute(PutCcrRestoreSessionAction.INSTANCE,
+                new PutCcrRestoreSessionRequest(sessionUUID, leaderShardId)).actionGet();
+            Store.MetadataSnapshot sourceFileMetaData = response.getStoreFileMetaData();
+            return new RestoreSession(remoteClient, sessionUUID, response.getNode(), indexShard, shardId, recoveryState,
+                sourceFileMetaData);
+        }
+
+        @SuppressWarnings("unchecked")
```

Comment:
why mark such a big method as unchecked? Isn't it possible to limit this annotation to a small part of the method?